### DISCUSSION TITLE
Break long class names into multiple lines? [Prettier Plugin]

### DISCUSSION THREAD

**Discussion by AndrejJurkin** [tailwindlabs/tailwindcss#7763] [Ideas]
Hey guys, it's not uncommon for class names to become very long, it's pretty hard to maintain and read. I somehow always break those class names into multiple lines manually, semantically even (by breakpoints). Are there any proper ways of dealing with this problem? In ideal case I could imagine being able to configure line breaks for classes in prettier plugin. Does it make any sense?

**Comment by bhison** [comment]
on the plus side, breaks you put in will not be changed at least, though I would love some line length enforcement, seems very strange that it's the only thing in my files which will expand over the max width

**Comment by rimzzlabs** [comment]
if you are using frontend framework like React.

the easiest way is by using [clsx](https://github.com/lukeed/clsx), you can separate the class by putting a comma, because you call a function that accept multiple arguments.

example on React:

```tsx
import clsx from "clsx"

export default function App(){
 return (
  <div className={clsx("w-full h-full", "flex items-center justify-center")}>
   // other components or 
  </div>
 )
}
```

optional:

Enhance it with [tailwind merge](https://github.com/dcastil/tailwind-merge) to solve style conflict, create a new util files like this one

```tsx
// @ utils/twclsx.ts

import clsx, { ClassValue } from "clsx"
import { twMerge } from "tailwind-merge"

export const twclsx(...args: ClassValue[]) => twMerge(clsx(...args))
```

**Reply by sookmax** [reply]
It would be really nice if I could write my tailwindcss classes like this in dev time only. 

And have them merged into one long complete string of classes at build time, so I don't have to worry about `clsx` being called every time my component renders at runtime.

**Reply by yunsii** [reply]
I have made a plugin [unplugin-polish-tagged-templates](https://github.com/yunsii/unplugin-polish-tagged-templates) to polish tagged templates, maybe a alternative method. here is a example:

```ts
// custom your tagged template function
function cls() {
  //...
}

// usage
const className = cls`
  cursor-pointer
  font-bold text-xl
  text-sky-500
  hover:text-sky-600
`
```

With the plugin, it will transfrom to:

```ts
const className = "cursor-pointer font-bold text-xl text-sky-500 hover:text-sky-600"
```

**Comment by Florian-crg** [comment]
You could use @emotion/css or styled-components or even const using the template literals principals

**Comment by danieluhl** [comment]
Would love addition to prettier for this, it's especially painful in vim/neovim where horizontal scrolling is rough, I manually do it like so:
```jsx
      <main
        className={`
            flex 
            h-screen 
            w-screen
            flex-col
            items-center
            justify-center 
            bg-gray-800
          `}
      >
```

**Reply by bradley** [reply]
This solution works for development if that's what youre after, but it also causes classes to have newlines in them in the web browser, which makes debugging from dev console very ugly.

**Reply by cyrus01337** [reply]
I think this would be nice if you could group utility classes based on what they're affecting and the modifier type/count, similar to what [`prettier-plugin-tailwindcss`](https://www.npmjs.com/package/prettier-plugin-tailwindcss) does.

**Reply by sgwrangler** [reply]
This will make my component files 3000 lines each

**Comment by Andrew-Pynch** [comment]
Is there any desire for something like this to get merged into tailwind? This would be a really exciting feature for me and my team to take advantage of

**Comment by dionysiusmarquis** [comment]
Currently I'm running a self written "minifier" using [node-html-parser](https://github.com/taoqf/node-html-parser) and [magic-string](https://github.com/rich-harris/magic-string). The minifier removes all newlines and unnecessary whitespace in any inline css  class declarations at preprocessor/bundler time. Would be nice to see something similar as an official tailwind option.

Here running as svelte preprocessor e.g.:
```js
import parser from "node-html-parser";
import MagicString from "magic-string";

/** @type {import('node-html-parser').default} */
const parse = parser.parse;

const regEx = /\s{2,}|[\n\r]+/g;

/** @type {(search: string) => RegExp} */
function createRegEx(search) {
  return new RegExp(search.replace(/[.*+\-?^${}()|[\]\\]/g, "\\$&"), "g");
}

/** @type {(node: Node, changes: string[][]) => void} */
function traverse(node, changes) {
  const childNodes = node.childNodes;
  if (childNodes && childNodes.length) {
    for (let i = 0; i < childNodes.length; i++) {
      /** @type {HTMLElement & { rawAttrs?: string }} */
      const childNode = childNodes[i];
      if (
        childNode.rawAttrs &&
        childNode.rawAttrs.match &&
        childNode.rawAttrs.match(regEx)
      ) {
        changes.push([
          childNode.rawAttrs,
          childNode.rawAttrs.replace(regEx, " "),
        ]);
      }
      traverse(childNode, changes);
    }
  }
}

/** @type {(filename: string, content: string, changes: string[][]) => { code: string map: string }} */
function applyChanges(filename, content, changes) {
  const s = new MagicString(content);
  for (let i = 0; i < changes.length; i++) {
    const [source, replace] = changes[i];
    const regex = createRegEx(source);
    const [...result] = content.matchAll(regex);
    for (let j = 0; j < result.length; j++) {
      const match = result[j];
      if (!isNaN(match.index)) {
        s.overwrite(match.index, match.index + source.length, replace);
      }
    }
  }
  var map = s.generateMap({
    source: filename,
    file: `${filename}.map`,
    includeContent: true,
  });

  return {
    code: s.toString(),
    map,
  };
}

/** @type {ReturnType<import('svelte-preprocess').default>} */
export const minifyHtmlAttributes = {
  markup({ content, filename }) {
    let code = content;
    let map = undefined;
    try {
      const html = parse(content);
      const changes = [];
      traverse(html, changes);
      const result = applyChanges(filename, content, changes);
      if (result) {
        code = result.code;
        map = result.map;
      }
    } catch (error) {
      console.error(error);
    }
    return { code, map };
  },
};
```

**Comment by chbybnwr** [comment]
this would result prettier git diff! :rocket:

**Comment by jpenna** [comment]
It makes sense to add line breaks to the plugin for the following reasons:

1. It would follow the directive of Prettier's `print-width` parameter
2. [Re-ordering is against Prettier's principles](https://github.com/prettier/prettier/issues/10918#issuecomment-854705502), so they can respect line breaks without any issue. But the main reason of this plugin is to re-order, so it should handle line breaks in an opinionated way.
3. The current state is flawed. It respect line breaks until it re-order and move classes between lines. For example, after adding `ui-active:text-red-600` at the begining of the string:

![Screenshot from 2023-05-18 12-23-12](https://github.com/tailwindlabs/tailwindcss/assets/16005946/5270daa3-2285-4b5c-88e6-0aca274ca63d)
Prettier sorts it last and moves `ui-active:bg-slate-100` up:

![Screenshot from 2023-05-18 12-23-24](https://github.com/tailwindlabs/tailwindcss/assets/16005946/e6956d21-da3a-436b-ba57-30181fd87e26)

The expected result would be to group `ui-active` together at the bottom line and to not split them, like so:

![Screenshot from 2023-05-18 12-22-52](https://github.com/tailwindlabs/tailwindcss/assets/16005946/1c9f5115-98a4-4bce-ac7a-67c5e73bcdea)

**Comment by dev-xo** [comment]
Any news on this? 
Would be really useful to have something that could avoid infinite class-names lines.

**Comment by code-creativeapps** [comment]
would definitely be useful

**Comment by CodeNinjaMatrix** [comment]
It would be really helpful if there is an option.

**Comment by renatoastra** [comment]
It would be very useful, any news?

**Comment by huggge** [comment]
Would be really nice

**Comment by bilalmalkoc** [comment]
Waiting for this feature.

**Comment by vlrevolution** [comment]
Waiting for this  feature as well

**Comment by blowsie** [comment]
More demand for the same topic
https://github.com/tailwindlabs/tailwindcss/discussions/11400
https://github.com/tailwindlabs/tailwindcss/discussions/11514

**Comment by blowsie** [comment]
https://github.com/tailwindlabs/prettier-plugin-tailwindcss/issues/39

**Comment by code-creativeapps** [comment]
Tailwind looses a bit of it's utility when the classes you add are hard to read when you come back to them or when someone else reads them.

So formatting the classes with line breaks, and also ordering by media queries like so would be really great :
```
mt-[10px] flex flex-wrap justify-center gap-x-[30px] gap-y-[20px] 
md:flex md:flex-row 
md:space-x-[20px] md:gap-x-0
lg:space-x-[40px]
```

**Reply by phendan** [reply]
This is *exactly* what I've been wishing for for ages

**Reply by harishash** [reply]
Thisss >>>>>>>

**Reply by theLeroy** [reply]
pls!

**Reply by willhindson** [reply]
Yes this is needed

**Reply by borela** [reply]
This plugin might help you:
https://github.com/borela-tech/multiline-tailwindcss/tree/main/packages/vite-plugin-multiline-tailwindcss

**Comment by tuffstuff9** [comment]
This would be an excellent addition! Currently, prettier's wrapping doesn't affect the classNames, at least with my current settings.

**Comment by yunsii** [comment]
I have made a plugin [unplugin-polish-tagged-templates](https://github.com/yunsii/unplugin-polish-tagged-templates) to polish tagged templates, maybe a alternative method. here is a example:

```ts
// custom your tagged template function
function cls() {
  //...
}

// usage
const className = cls`
  cursor-pointer
  font-bold text-xl
  text-sky-500
  hover:text-sky-600
`
```

With the plugin, it will transfrom to:

```ts
const className = "cursor-pointer font-bold text-xl text-sky-500 hover:text-sky-600"
```

It required a prettier/eslint plugin to **format internal strings and expressions** if the tagged template used **in deep indent**. If not a proper plugin to format, I will try to create one laterly...

---

After deeper investigation, prettier-plugin-tailwindcss not works well if indent change:

```tsx
const Component = (
  <p
    className={cls`
      text-sm
    `}}
  >
    Hi
  </p>
)
```

After add a dom to wrap `p` tag, it still indent with 4 spaces.

**Comment by yunsii** [comment]
After more deeper development and investigation, I think I find a best solution for now.

- [tagged-classnames-free](https://github.com/yunsii/tagged-classnames-free) for write whatever you want classnames
- [eslint-plugin-unicorn](https://github.com/sindresorhus/eslint-plugin-unicorn)'s [template-indent](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/template-indent.md) rule to auto indent/dedent
- [unplugin-polish-tagged-templates](https://github.com/yunsii/unplugin-polish-tagged-templates) to optimize runtime performance

Try overall features on [Gitpod](https://gitpod.io/#https://github.com/yunsii/tagged-classnames-free).

**Comment by SebiBasti** [comment]
I'm also interested in this!

**Comment by nOji** [comment]
This is one of the main downsides of working with TailwindCSS in modern web development.

**Comment by Andrew-Pynch** [comment]
has anyone found a good way to integrate these proposed sollutions into neovim?

**Reply by Andrew-Pynch** [reply]
At this rate I'm gonna have to do the unthinkable and ask my boss if I can contribute to open source to make this myself 😆

**Comment by pyropecs** [comment]
Instead of doing this long class names. We can just put that in json file and we can parse it as an object and add it to as string. This enables the our own naming conventions into the bunch of class names

**Comment by phendan** [comment]
Just to add another voice, having the option to format classes over multiple lines would effectively eliminate the one downside (and initial hurdle for many newcomers) that Tailwind has: the line length. As others have mentioned it goes against common formatting conventions to leave the length unrestricted, but it's not just an aesthetic thing. It also makes the styles less legible compared to the ideas cited in this thread, especially the ones where the classes are also grouped, for example by breakpoint. Even though I love Tailwind enough to use it on basically every project, and have been for some time, the fact that this isn't a thing yet *still* rubs me the wrong way and is always at the back of my mind. This plugin seems like the perfect place to add this functionality.

**Comment by Diankes** [comment]
This is not integrated in the tailwind plugin to date, currently the best solution is still shift + scroll and a UW monitor or eyeballing it with template strings.

**Comment by ony3000** [comment]
Hello everyone!

There is a [prettier plugin](https://www.npmjs.com/package/prettier-plugin-classnames) that I released this year. Maybe this will help.

**Reply by kaladivo** [reply]
This works very well, thanks :)!

**Reply by BrandenXia** [reply]
It works great for me :)

**Reply by mak33v** [reply]
Finally it happened, thank you very much for the plugin.

**Reply by sheeyphone** [reply]
That's really helpful! Thanks!

**Reply by xanderios** [reply]
It works just fine, thanks :)

**Reply by thype-it** [reply]
Hi, love the plugin but i experienced an issue. It seems not to recognize componets in this format <RootComponent.Component> and therefore does not work in those cases.

**Reply by ozanyusufoglu** [reply]
Works great, thanks!

**Reply by ony3000** [reply]
@thype-it Fixed in `v0.6.4`, check it out!

**Reply by SebiBasti** [reply]
Works like a charm!

**Reply by Velkamhell97** [reply]
It works with html default class instead of className ?

**Reply by ony3000** [reply]
@Velkamhell97 When I first created this plugin, I missed frameworks like `Solid`. Therefore, the `customAttributes: ['class']` option must be added to `*.js(x)` or `*.ts(x)` files for it to work.

`*.html` files are not yet supported and are planned to be included in the next minor release.

**Reply by monecchi** [reply]
I’ll give it a try for sure! Thank you!

**Reply by realChakrawarti** [reply]
@ony3000 👏🏼

**Reply by DerekCL** [reply]
@ony3000 thanks for creating this. Really excited for the *.html files to be supported!

**Reply by vanngoh** [reply]
@ony3000 Does this also support `.vue` file? 🤩

**Reply by ony3000** [reply]
@vanngoh Yes, `*.vue` files are supported! :raised_hands:

(You may have already read [this guide](https://github.com/ony3000/prettier-plugin-classnames/issues/33). :joy:)

**Reply by ony3000** [reply]
@Velkamhell97 `v0.7.0` has been released. The `class` attribute is now also supported by default!

**Reply by agestasaputra** [reply]
Hey everyone, can you guys share the docs how to implement `eslint-plugin-readable-tailwind`  into eslint flat config?
Thank you

**Reply by ony3000** [reply]
@agestasaputra Please refer to [these documents](https://github.com/schoero/eslint-plugin-readable-tailwind/tree/main/docs/parsers). And I think you've come to the wrong thread. :joy:

**Comment by letoast** [comment]
https://www.npmjs.com/package/eslint-plugin-readable-tailwind

**Reply by devrivers-high** [reply]
So far `eslint-plugin-readable-tailwind@1.8.1`  plugin provides full flexibility on the config, I was able to achieve 1 line per class name using template string literal and sorted by recommended class order.

The [Prettier community says the opposite](https://github.com/prettier/prettier/issues/10918#issuecomment-971641639), but that's why we have plugins to customize our needs.

I tested the config using ViteJS with tailwind css setup.

From this hard to maintain code
```tsx
export function Button() {
  return (
    <button
      type="button"
      className="rounded bg-indigo-600 px-2 py-1 text-xs font-semibold text-white shadow-sm hover:bg-indigo-500 focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 focus-visible:outline-indigo-600"
    >
      Click Me
    </button>
  );
}

```


To this easy to maintain and debug code:
```tsx
 export function Button() {
  return (
    <button
      type="button"
      className={`
        inline-flex
        h-10
        min-w-12
        items-center
        gap-x-1.5
        rounded-md
        bg-black
        px-2.5
        py-1.5
        text-sm
        font-semibold
        text-white
        shadow-sm

        focus-visible:outline
        focus-visible:outline-2
        focus-visible:outline-offset-2
        focus-visible:outline-indigo-600

        hover:bg-indigo-500aaa
      `}
    >
      Click Me
    </button>
  );
}
```


eslint config:
```js
// eslint.config.js
export default tseslint.config(
  {
    /* your config goes here */
    plugins: {
      /* your plugins goes here */
      // enable plugin
      "readable-tailwind": eslintPluginReadableTailwind,
    },
    rules: {
      /* your rules goes here */
      // enable all recommended rules to warn
      ...eslintPluginReadableTailwind.configs.warning.rules,
      // enable all recommended rules to error
      ...eslintPluginReadableTailwind.configs.error.rules,

      // or configure rules individually
      "readable-tailwind/multiline": ["warn", { classesPerLine: 1 }],
    },
  }
);
```

It also works with variables (ex:. ```const className = `...` ```), but the tailwind intellisense is not as smart to do the job within a variable, so I keep the class names within the HTML tags for now.

**Reply by Redmega** [reply]
fyi you can customize the tailwind intellisense setting `tailwindCSS.experimental.classRegex` to customize that. I use that to get intellisense within `clsx()` and `tailwind-variants` calls.

You can probably use something like this for the variables:
https://github.com/paolotiu/tailwind-intellisense-regex-list#javascript-string-variable

**Reply by markuskullberg-sanoma** [reply]
The newlines and extra whitespace end up in the rendered markup as well, right? Did you do something about it or don't care?

**Reply by letoast** [reply]
No, they shouldn't. It might depend on the build tool, but in my case (i use Nuxt and it uses vite by default) the extra newlines and whitespace don't end up in the rendered markup.

**Comment by griga** [comment]
sadly non  of this options have worked for me. 

my vscode  solution is to quickly toggle _word wrapping_ in file with _alt+z_
+ also set _Wrapping Indent_ setting to _indent_

**Comment by cupid20103** [comment]
There are some way to solve this problem.
You can solve this problem by using tailwind-css and styled-components together.
You can refer it on this: [Tailwind-Styled-Component](https://www.npmjs.com/package/tailwind-styled-components)

I hope this helps you.

**Comment by Ayfri** [comment]
Any update ?

**Comment by Redmega** [comment]
Would love to use a template string for multiline. I think a more nuanced whitespace setting would be good. RIght now it's too strict, if it could strip only spaces but leave newlines alone it would allow for things like

```tsx
<div className={`
	tw-inline-flex tw-items-center tw-justify-center tw-gap-2 tw-border-2 tw-border-transparent tw-text-center
	tw-font-body tw-font-semibold tw-transition hover:tw-text-current hover:tw-shadow-active disabled:tw-shadow-none
`} />
```

**Comment by BatthewZ** [comment]
Just gonna leave these here:
https://www.npmjs.com/package/prettier-plugin-classnames
https://www.npmjs.com/package/eslint-plugin-readable-tailwind

**Reply by letoast** [reply]
Both of these solutions were already posted above

**Reply by Ayfri** [reply]
We tried the first plugin and it wasn't doing anything sadly

**Reply by ony3000** [reply]
> it wasn't doing anything

@Ayfri Could you elaborate a little more on what this means?

I'm wondering if the line wrapping either didn't work, or if they did work but didn't produce the results you expected.

**Reply by Ayfri** [reply]
They didn't work at all, it was not changing anything to my code.

**Reply by ony3000** [reply]
My guess is that it's either a bug in prettier-plugin-classnames (which usually outputs `The second format failed.`) or something is missing in your Prettier configuration.

Have you checked out [this guide](https://github.com/ony3000/prettier-plugin-classnames/issues/33)?

**Reply by Ayfri** [reply]
Yes, that's how I've installed it

**Reply by ony3000** [reply]
Then may be it's a bug in prettier-plugin-classnames.

Please report detail information (configuration, code snippet, public repository, etc.) that would allow me to reproduce the issue to [here](https://github.com/ony3000/prettier-plugin-classnames/issues) and I'll check it out.

**Comment by wazeerc** [comment]
Still nothing? :o

**Reply by wazeerc** [reply]
This worked for me: https://www.npmjs.com/package/eslint-plugin-readable-tailwind

**Comment by Seraphim-AT** [comment]
Hi folks! What about this approach:

```
<div className={[
    "absolute",
    "w-[24rem]",
    `top-[${value}rem]`,
    "h-[24rem]"
].join(" ")}>
```

**Reply by lirc571** [reply]
Does the tailwind VS Code extension (and other editors / IDEs) work with this approach (e.g. auto-complete, doc)?

**Reply by borela** [reply]
This plugin might help you:
https://github.com/borela-tech/multiline-tailwindcss/tree/main/packages/vite-plugin-multiline-tailwindcss

**Comment by Seraphim-AT** [comment]
@lirc571 I'm working in WebStorm, and it works fine there

**Comment by wazeerc** [comment]
Ngl I stopped caring about this issue and started using linters with plugins instead - eslint or biome.

**Comment by borela** [comment]
In case anyone still wants this, I created a [vite](https://github.com/borela-tech/multiline-tailwindcss/tree/main/packages/vite-plugin-multiline-tailwindcss) plugin that allows classes to be broken across multiple lines.

```html
<div className="
  bg-[
    linear-gradient(
      to_right,
      theme(colors.purple.600),
      theme(colors.purple.900),
    ),
  ]
">
</div>

<!-- Becomes -->
<div className="bg-[linear-gradient(to_right,theme(colors.purple.600),theme(colors.purple.900))]">
</div>
```

and

```js
const BODY_CSS = tailwindcss`
  bg-[
    linear-gradient(
      to_right,
      theme(colors.purple.600),
      theme(colors.purple.900),
    ),
  ]
`

// Becomes
const BODY_CSS = `bg-[linear-gradient(to_right,theme(colors.purple.600),theme(colors.purple.900))]`

**Comment by aria-amini** [comment]
An unconventional option that has been working for me is to have vscode wrap lines that exceed the print width using this option:
"editor.wordWrap": "wordWrapColumn"

**Reply by borela** [reply]
Git diffs will still be painful.
